pheget/api/__init__.py

```diff
@@ -2,13 +2,73 @@
 API endpoints (return JSON, not HTML)
 """
 
+import json
+import sqlite3
+import urllib.request
+
 from flask import Blueprint, abort, jsonify, request
 
+from .. import model
 from .format import query_variants
 
 api_blueprint = Blueprint("api", __name__)
 
 
+@api_blueprint.route("/bestvar/<string:symbol>/", methods=["GET"])
+def gene_query(symbol: str):
+    """
+    Given a gene, query an SQL database to find the variant with the strongest association with that gene
+    """
+    # ~UGLY HACK~ If omnisearch did not return a gene_id, then searchbox will fill in "undefined" - immediately abort in that case
+    if symbol == "undefined":
+        abort(404)
+    # Instead of querying locally, I will use omnisearch to get the relevant parameters (chrom, pos, gene_id)
+    url = f"https://portaldev.sph.umich.edu/api/v1/annotation/omnisearch/?q={symbol}&build=GRCh38"
```

Comment:
Interesting. Does the backend need to talk to omnisearch, or could this be handled by the frontend? (I am partly asking because this endpoint seems to be triggered in response to an omnisearch call; maybe I am misreading that)

https://github.com/statgen/pheget/pull/47/files#diff-fb9be9e8a12a6fe4420134c94525c670R70-R82

Reply:
This one was a somewhat unfortunate compromise I made in designing the API: 
I wanted it to be usable directly (that is, it will return the best variant given just the name of a gene). Unfortunately, to make the performance of the lookup acceptable requires not just gene_id, but also chrom, start, and end. Since I am not storing the full mapping of gene_ids to chrom:start-end (maybe I shoiuld?), I'm using omnisearch to retrieve those fields.

As you noticed, the searchbox's parsing of the search query _also_ looks up omnisearch, so currently we have the unfortunate double-lookup every time you search a field that isn't chrom:pos, chrom:start-end, or rs#: once to figure out if the query is a gene in searchbox.js, and once to retrieve the chrom/start/end in the API in __init__.

I could change the behavior so that if searchbox hits omnisearch once, it passes chrom/start/end to the API, which would prevent the double-hit.

Reply:
I'd need to sit and think on this one a bit; nudge me tomorrow? If we can do this without the double request that would be ideal, but it's not the kind of question I should attempt to answer without thought and care.

One of the less obvious ways this can get weird: if the external service is slow to respond it could tie up some server resources in the synchronous python request/ response cycle, which would let a few people hitting our server tie up pheget (if omnisearch is slow to respond, then all our workers would be in use waiting for it). Imagine one person with a full cart of groceries at the front of the "10 items or less" line. Not fun.

There are some deployment changes we can make, like ensuring the use of async workers in gunicorn, that can ensure that one long synchronous request would not prevent others from being serviced. 

But if we can do this in fewer requests, and all in browser, that's certainly ideal. :)